fix: SelectFields should update options when props change

bdow opened the pull request:
When the 'options' prop changes, then update the internal state.
Adds prop for displaying loading indicator for cases where the props are lazily loaded into the menu.

## Perf Test Before
<img width="1038" alt="Before" src="https://user-images.githubusercontent.com/1143861/147285827-f48a3b07-f316-43f9-ab93-dac7131e8dbd.png">

## Perf Test After
<img width="1031" alt="After" src="https://user-images.githubusercontent.com/1143861/147285843-61eabeb9-140b-45db-917e-8c6f2b0ff627.png">

stephenh reviewed (approved):
Maybe add a story for the new loading UI? ...or does that require focusing into the element? FWIW the latest storybook has the `play` feature that means we should be able to start "click / focus an element" before the snapshots run. Granted, don't have to tackle that for this PR.

I think eventually I'd like `optionsLoading` to be driven from a promise that comes from an `option={() => Promise<...>}` sort of setup, but this does what we need for now so sgtm.

bdow commented:
> Maybe add a story for the new loading UI? ...or does that require focusing into the element? FWIW the latest storybook has the `play` feature that means we should be able to start "click / focus an element" before the snapshots run. Granted, don't have to tackle that for this PR.
> 
> I think eventually I'd like `optionsLoading` to be driven from a promise that comes from an `option={() => Promise<...>}` sort of setup, but this does what we need for now so sgtm.

@stephenh - I have changed this up quite a bit. `options` now can be a tuple with an initial array of options and a load callback that, for now, returns an updated list of options (meant to replace the initial list). This removed the need for an `optionsLoading` prop as we can now handle that state locally. 
I added tests to ensure options can be updated either via the Promise, or just by a change of the options prop. I also added a Storybook story for the loading state of the options utilizing the new Storybook 'play' feature as you suggested. 
I've tested this locally with `internal-frontend` and it all is working as expected. No rush on a second review as I don't plan on merging until after break.

bdow commented:
@stephenh would you mind looking at this again? It has changed a lot since you approved.

KoltonG commented on src/inputs/SelectField.stories.tsx:
```diff
@@ -197,6 +198,61 @@ export const Contrast = Template.bind({});
+LoadingState.parameters = {
+  chromatic: { delay: 1000 },
+};
+LoadingState.play = async ({ canvasElement }: { canvasElement: HTMLElement }) => {
```
**suggestion:** It would be helpful if you could add a comment on what this does! 

Just checked the the Storybook and see that there are alignment weirdness 

<img width="678" alt="CleanShot 2022-01-10 at 14 08 23@2x" src="https://user-images.githubusercontent.com/3606121/148816815-0f0d901b-2dee-48ee-9f7b-45c07fb8fd36.png">

bdow replied:
Yea, the alignment is off a bit when the field is flush on the edge of the screen. Not sure why, I think its a React-Aria thing, but haven't dug into it.

KoltonG commented on src/inputs/SelectField.test.tsx:
```diff
@@ -154,17 +149,74 @@ describe("SelectFieldTest", () => {
+    // Then expect to see the initial option
+    expect(r.getAllByRole("option")).toHaveLength(1);
+    // And when changing the options
+    click(r.updateOptions);
```
Ouu how does this work?

bdow replied:
Not 100% what you're referring to when you ask this. Basically, I put a button in the `TestSelectField` component defined in the test that would change the `options` prop value.

KoltonG commented on src/components/CssReset.tsx:
```diff
@@ -14,8 +14,35 @@ export function CssReset() {
+  /**
+   * Beam animations
+   */
+  @keyframes loadingDots {
```
Ahh got it!

KoltonG reviewed (approved):
🍣 super awesome work Brandon! Lots here that I didn't have time to get into but looks exciting from what I could understand!

bdow commented on package.json:
```diff
@@ -71,6 +71,7 @@
   "devDependencies": {
     "@babel/core": "^7.15.5",
     "@babel/plugin-proposal-class-properties": "^7.14.5",
+    "@babel/preset-env": "^7.16.5",
```
Not new, but we had some missing dependencies which were being included via another package. Though, it seemed that updating storybook called out some of these missing hard dependencies.

bdow commented on src/utils/index.test.ts:
```diff
@@ -0,0 +1,20 @@
+      expect(areArraysEqual([1, 2, 3], [1, 2, 3])).toBe(true);
+      expect(areArraysEqual([], [])).toBe(true);
+      expect(areArraysEqual([1], [2, 3])).toBe(false);
+      expect(areArraysEqual([1, 2, 3], [1, 3, 2])).toBe(false);
```
Correct.

bdow commented on src/inputs/SelectField.tsx:
```diff
@@ -43,10 +39,9 @@ export function SelectField<O, V extends Value>(
-          onSelect && selectedOption && onSelect(getOptionValue(selectedOption), selectedOption);
+      onSelect={(values, options) => {
+        if (values.length > 0 && options.length > 0) {
+          onSelect && onSelect(values[0], options[0]);
```
We do. We could update to that.

bdow commented on src/inputs/SelectField.tsx:
```diff
@@ -43,10 +39,9 @@ export function SelectField<O, V extends Value>(
-          const selectedOption = options.find((o) => getOptionValue(o) === values[0]);
-          onSelect && selectedOption && onSelect(getOptionValue(selectedOption), selectedOption);
+      onSelect={(values, options) => {
+        if (values.length > 0 && options.length > 0) {
```
Hmm... I think it was a TypeScript thing here.

bdow commented on src/inputs/internal/LoadingDots.tsx:
```diff
@@ -0,0 +1,35 @@
+import { Css, Palette } from "src/Css";
+import { useTestIds } from "src/utils";
+
+export function LoadingDots({ contrast }: { contrast: boolean }) {
```
Right now for SelectFields, though we could potentially expand it to other places if needed.

bdow commented on src/inputs/internal/LoadingDots.tsx:
```diff
@@ -0,0 +1,35 @@
+  const circleCss = Css.hPx(8)
+    .wPx(8)
+    .br4.bgColor(contrast ? Palette.Gray500 : Palette.Gray300)
+    .add("animationName", contrast ? "loadingDotsContrast" : "loadingDots")
```
Yes, they are in the CSS Reset that ships with Beam

bdow commented on src/inputs/internal/SelectFieldInput.tsx:
```diff
@@ -142,35 +125,36 @@ export function SelectFieldInput<O, V extends Value>(props: SelectFieldInputProp
-            state.selectionManager.setSelectedKeys(state.selectionManager.selectedKeys);
+            // Always call `resetField` onBlur, this ensures the field's `input.value` resets
+            // to what it should be in case it doesn't currently match.
+            resetField();
```
The external caller won't need to be aware of this. This explicitly to reset the field's internal state. For instance, if I am typing to filter the options and then I blur without changing an option, it will reset to its previous value.

bdow commented on src/inputs/internal/SelectFieldBase.tsx:
```diff
@@ -10,25 +10,41 @@ import { ListBox } from "src/inputs/internal/ListBox";
+  onSelect: (values: V[], opts: O[]) => void;
   multiselect?: boolean;
+  disabledOptions?: V[];
+  options: O[] | { initial: O[]; load: () => Promise<{ options: O[] }> };
```
The Project Switcher is one case where we do something like:
**In Blueprint** - Request the project information required for displaying the current project in the Project Switcher. This will be an API request of only one option (the current project). This is what is populated in the `initial` prop. Once the user interacts with the SelectField, then the `load` event is fired. Handling the promise locally in the SelectField allows us to manage the loading state internally, vs requiring a `optionsLoading: boolean;` prop to be changed by the caller.

bdow commented on src/inputs/internal/SelectFieldBase.tsx:
```diff
@@ -133,41 +171,49 @@ export function SelectFieldBase<O, V extends Value>(props: SelectFieldBaseProps<
+  const firstOpen = useRef(true);
   function onOpenChange(isOpen: boolean) {
+    if (firstOpen.current && isOpen) {
+      maybeInitLoad();
```
Yes, we could immediately load the options, but we my thinking was to reduce the number of API calls until it is actually required.

bdow replied:
We do not reset `firstOpen` when the user sends in new options at the moment. The expectation is that the user would follow the pattern of `load` being the final step to load all options.
